tsinfer/inference.py

```diff
@@ -714,34 +715,41 @@ def match_ancestors_batch_init(
     for group_index, group_ancestors in matcher.group_by_linesweep().items():
         # Make ancestor_ids JSON serialisable
         group_ancestors = list(map(int, group_ancestors))
-        partitions = []
-        current_partition = []
-        current_partition_work = 0
-        # TODO: Can do better here by packing ancestors
-        # into as equal sized partitions as possible
+        # The first group is trivial so never partition
         if group_index == 0:
-            partitions.append(group_ancestors)
+            partitions = [
+                group_ancestors,
+            ]
         else:
             total_work = sum(ancestor_lengths[ancestor] for ancestor in group_ancestors)
-            min_work_per_job_group = min_work_per_job
-            if total_work / max_num_partitions > min_work_per_job:
-                min_work_per_job_group = total_work / max_num_partitions
-            for ancestor in group_ancestors:
-                if (
-                    current_partition_work + ancestor_lengths[ancestor]
-                    > min_work_per_job_group
-                ):
-                    partitions.append(current_partition)
-                    current_partition = [ancestor]
-                    current_partition_work = ancestor_lengths[ancestor]
-                else:
-                    current_partition.append(ancestor)
-                    current_partition_work += ancestor_lengths[ancestor]
-            partitions.append(current_partition)
+            parition_count = math.ceil(total_work / min_work_per_job)
+            if parition_count > max_num_partitions:
+                parition_count = max_num_partitions
+
+            # Partition into roughly equal sized bins (by work)
+            sorted_ancestors = sorted(
+                group_ancestors, key=lambda x: ancestor_lengths[x], reverse=True
+            )
+            partitions = []
+            partition_lengths = []
+            for _ in range(parition_count):
+                partitions.append([])
+                partition_lengths.append(0)
+
+            # Use greedy bin packing - place each ancestor in the bin with
+            # lowest total length
+            for ancestor in sorted_ancestors:
```

Comment:
How about we use a heapq for this?
```python
heap = [(0, []) for _ range(partition_count]
for ancestor in sorted_ancestors:
      sum_len, partition = heapq.heappop(heap)
      partition.append(ancestor)
      sum_len += ancestor_lengths[ancestor]
      heapq.heappush(heap, (sum_len, partition))
```
I think this does the same thing, but avoids the quadratic time complexity here.

Reply:
Very nice, I should have though of this!